### ISSUE TITLE
Node runs out of memory during production build

### ISSUE THREAD

**Issue by trentmwillis** [ember-cli/stress-app#14]
From a clean clone, running `ember build --environment=production` yields the following error:

```bash
<--- Last few GCs --->

  151400 ms: Scavenge 1394.2 (1456.9) -> 1394.2 (1456.9) MB, 0.9 / 0 ms (+ 5.6 ms in 1 steps since last GC) [allocation failure] [incremental marking delaying mark-sweep].
  152541 ms: Mark-sweep 1394.2 (1456.9) -> 1254.8 (1456.9) MB, 1141.6 / 0 ms (+ 9.9 ms in 3 steps since start of marking, biggest step 5.6 ms) [last resort gc].
  154054 ms: Mark-sweep 1254.8 (1456.9) -> 1254.7 (1456.9) MB, 1512.8 / 0 ms [last resort gc].


<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x2259336e3ac1 <JS Object>
    1: has_side_effects [0x225933604189 <undefined>:~5889] [pc=0x328e06438e03] (this=0x38dfe0a110e9 <an AST_New with map 0x3177f454aa39>,compressor=0x2f16a43053c9 <a Compressor with map 0x3177f45b57b1>)
    2: /* anonymous */(aka /* anonymous */) [0x225933604189 <undefined>:6042] [pc=0x328e065f9b07] (this=0x225933604189 <undefined>,def=0x2851bf7f59c1 <an AST_VarDef with map 0x3177f45afcf1>)
   ...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - process out of memory
Abort trap: 6
```

However, a development `ember build` completes rather quickly without any issues.

We have been seeing this issue occur with our large application as well, which is why I cloned the stress-app in the first place. My gut instinct is that the Uglify process (which seems to be one of the few diffs between development and production builds) doesn't handle exceedingly large (> 8 mb or so) JS files.

Here's some relevant environment info as well:

```bash
ember --version
ember-cli: 2.5.0
node: 5.4.0
os: darwin x64
```

cc @chadhietala

**Comment by stefanpenner** [comment]
@trentmwillis ya this uglify,

**Comment by trentmwillis** [comment]
@stefanpenner any tips on making them happy? Aside from just increasing the memory limit, my first thought is to split up the large files, but that seems like it would require custom build tooling.

**Comment by chadhietala** [comment]
@trentmwillis not saying this should be the solution but can try running this on Node 6?

**Comment by stefanpenner** [comment]
@trentmwillis some ideas:

* split the files (likely best win all around, something via engines)
* make sure we aren't minifying something already minified
* some uglify settings are worse then others.
* shell out to google clojure compiler
* Do some magic to split behind the scenes, and uglify a file in two chunks.
* We should likely investigate if uglify can be improved (I'm sure it wouldn't hurt)

**Comment by stefanpenner** [comment]
It's good that this repo also demonstrates it, as we have something OSS to experiment with.

**Comment by trentmwillis** [comment]
@chadhietala same issue when using Node 6.

@stefanpenner I tend to think splitting the files and encouraging solutions around Engines/distributed code makes the most sense as well. Since this issue is only really likely to occur in large, monolithic codebases, I feel that the best thing we can do is encourage teams to break up those monoliths.

That said, I'll still probably look into Uglify a bit more and see if gains can be made there that will help across the board.

**Comment by stefanpenner** [comment]
@trentmwillis our thoughts are aligned entirely.

**Comment by xdumaine** [comment]
@trentmwillis @stefanpenner do yall know if there is anything specifically tracked around uglify? We have an enormous ember application, enormous even after we've broken out 7 different engines. It uses upwards of 3gb of memory for a production build. We're open for sharing some data/output from our builds if that would be helpful. cc @toddjordan

**Comment by stefanpenner** [comment]
> do yall know if there is anything specifically tracked around uglify?

I'm not sure any work is happening on the uglify side :(  It should be possible to make it less memory hungry, but I'm not sure..

Alternatively you could uglify each module, then concat the uglified modules. Which will likely be much faster, and only slightly larger.

**Comment by trentmwillis** [comment]
Nothing is happening as far as I know. I also doubt it is likely to gain any traction since "best practices" would be to ship smaller files to your clients.

**Comment by stefanpenner** [comment]
@trentmwillis nah, I think we will look to address.

Most likely via:

> Alternatively you could uglify each module, then concat the uglified modules. Which will be much faster (and could run in parallel), and only slightly larger output.

**Comment by xdumaine** [comment]
Thanks. Our use case is a less common one since it's a thick app by nature, with an O(1) users on the app for a long time, so while we still strive for fast and small, that's a lower priority to the completeness of the app. We've been increasing CI to do production builds and deploys of branches, which is how we've begun getting burned by memory consumption.

**Comment by stefanpenner** [comment]
> Our use case is a less common one since it's a thick app by natur

Nah this is extremely common. We should be sure to support it nicer.

**Comment by xdumaine** [comment]
This has grown for us as the app has grown. We're at upwards of 4.5 gb of memory now, and our vendor.js is 21 mb minified (~46 mb unminified). We've been doing some investigation into that and found that (because we leverage engines and addons heavily) we're getting a lot of dependencies duplicated.  These libs were included in vendor.js over 20 times (the number at the end is the times it was defined in vendor.js). 

![image](https://user-images.githubusercontent.com/833911/28037442-47f7b39e-6589-11e7-8e8d-9e1807de7619.png)

**Comment by xdumaine** [comment]
So we have a shared common components addon (I'll refer to that as EPC). The parent app uses EPC directly for some components, and it also uses 13 other engines/addons, most of which also use EPC directly for some components. So during a build, our app loads up each addon and runs its included. Part of that is going through the addon's children and also running included, to bring in the dependencies for those as well. Because EPC is a dependency of each of our addons/engines, it gets included n times, where n is the number of addons/engines that include EPC as a dependency. So how all those deps were included twice for each engine, I do not know. Lodash was included 40+ times, and we managed to reduce that.

**Comment by xdumaine** [comment]
Possibly related https://github.com/ember-engines/ember-engines/issues/405 https://github.com/ember-engines/ember-engines/issues/438
